velox/type/Decimal.h

```diff
@@ -0,0 +1,216 @@
+#include "velox/type/StringView.h"
+
+namespace facebook::velox {
+
+#define PRECISION(typmod) ((uint8_t)(typmod >> 8))
+#define SCALE(typmod) ((uint8_t)(typmod))
+#define TYPMOD(X, Y) ((X << 8) | y)
+#define NUM_DECIMAL_PARAMETERS 2
+
+using int128_t = __int128_t;
+static constexpr uint8_t kMaxPrecisionInt128 = 38;
+static constexpr uint8_t kDefaultScale = 0;
+static constexpr uint8_t kDefaultPrecision = kMaxPrecisionInt128;
+
+/*
+ * This class defines the Velox DECIMAL type support to store
+ * fixed-point rational numbers.
+ */
+class BigDecimal {
+ public:
+  inline const uint8_t getPrecision() const {
+    return precision_;
+  }
+
+  inline const uint8_t getScale() const {
+    return scale_;
+  }
+
+  inline int128_t getUnscaledValue() const {
+    return unscaledValue_;
+  }
+
+  inline void setUnscaledValue(const int128_t& value) {
+    unscaledValue_ = value;
+  }
+
+  // Needed for serialization of FlatVector<BigDecimal>
+  operator StringView() const {VELOX_NYI()}
+
+  std::string toString() const;
+
+  operator std::string() const {
+    return toString();
+  }
+
+  bool operator==(const BigDecimal& other) const {
+    return (
+        this->unscaledValue_ == other.getUnscaledValue() &&
+        this->precision_ == other.getPrecision() &&
+        this->scale_ == other.getScale());
+  }
+  bool operator!=(const BigDecimal& other) const {
+    VELOX_NYI();
+  }
+
+  bool operator<(const BigDecimal& other) const {
+    VELOX_NYI();
+  }
+
+  bool operator<=(const BigDecimal& other) const {
+    VELOX_NYI();
+  }
+
+  bool operator>(const BigDecimal& other) const {
+    VELOX_NYI();
+  }
+
+  BigDecimal(
+      int128_t value,
+      uint8_t precision = kDefaultPrecision,
+      uint8_t scale = kDefaultScale)
+      : unscaledValue_(value), precision_(precision), scale_(scale) {}
+  BigDecimal() = default;
+
+ private:
+  int128_t unscaledValue_; // The actual unscaled value with
+                           // max precision 38.
+  uint8_t precision_ = kDefaultPrecision; // The number of digits in unscaled
+                                          // decimal value
+  uint8_t scale_ = kDefaultScale; // The number of digits on the right
+                                  // of radix point.
+};
+
+class ShortDecimal {
+ public:
+  ShortDecimal(
+      int128_t value,
+      uint8_t precision = kDefaultPrecision,
+      uint8_t scale = kDefaultScale)
+      : unscaledValue_(value), precision_(precision), scale_(scale) {}
+
+  ShortDecimal() = default;
+
+ private:
+  int64_t unscaledValue_; // The actual unscaled value with
+                          // max precision 38.
+  uint8_t precision_ = kDefaultPrecision; // The number of digits in unscaled
+                                          // decimal value
+  uint8_t scale_ = kDefaultScale; // The number of digits on the right
+                                  // of radix point.
+};
+
+class DecimalCasts {
+ public:
+  static BigDecimal parseStringToDecimal(const std::string& value) {
```

Comment:
It looks like this logic comes from com.facebook.presto.common.type.Decimals#parse which is used to parse decimal literals found in SQL queries, e.g.

```
SELECT DECIMAL '1.234';
```

This logic is different from casting a string to decimal. The cast operation must specify the specific target type, e.g. decimal of a particular precision and scale.

```
SELECT cast('1.234' as DECIMAL(1, 4));
```

I don't think we need the logic of parsing decimal literals. Do we? Let's remove it from the PR to make it simpler.